File: algorithms/go/003_lengthOfLongestSubstring.go
package algorithms

func lengthOfLongestSubstring(s string) int {

	max := 0
	i := 0
	j := 0

	ht := make(map[byte]int)

	for i < len(s) && j < len(s) {
		if _, ok := ht[s[j]]; ok {
			delete(ht, s[i])
			i++
		} else {
			ht[s[j]]++
			j++
			if j-i > max {
				max = j - i
			}
		}
	}
	return max
}

File: algorithms/c/003_lengthOfLongestSubstring.c

#include <stdio.h>
#include <string.h>

int lengthOfLongestSubstring(char *s) {
    int ht[256] = {0};

    int max = 0;
    int i = 0;
    int j = 0;

    int len = strlen(s);

    while (i < len && j < len) {
        if (!ht[s[j]]) {
            ht[s[j]]++;
            j++;
            if (j - i > max) {
                max = j - i;
            }
        } else {
            ht[s[i]]--;
            i++;
        }
    }

    return max;
}

void test_lengthOfLongestSubstring() {
    char *s = "abcabcbb";
    int ret = lengthOfLongestSubstring(s);
    printf("3 = %d\n", ret);

    s = "bbbbb";
    ret = lengthOfLongestSubstring(s);
    printf("1 = %d\n", ret);

    s = "pwwkew";
    ret = lengthOfLongestSubstring(s);
    printf("3 = %d\n", ret);
}